drilldown/filters.py

```diff
@@ -0,0 +1,157 @@
+           inserted in the phrase "Showing voters who <Description>".
+           E.g. "are <em>female</em>", "live in <em>Orange</em> county",
+           "have an <em>active</em> registration".
+           By convention we put ``em`` around the part that can change.
+    """
+    editing_template = Template("""
+        <select name="{{ filter.field_name }}">
+          <option>-</option>
+          {% for value, label, foo in filter.choices %}
+            <option value="{{ value }}" {% if value in filter.values %}selected{% endif %}>{{ label }}</option><br/>
+          {% endfor %}
+        </select>
+    """)
+
+    def __init__(self, display_name: str, field_name: str, choices: List[Tuple]):
+        super().__init__(display_name, field_name)
+        self.choices = choices
+        if not choices:
+            raise ValueError("Choices must not be empty")
+
+        for choice in choices:
+            if len(choice) != 3:
+                raise ValueError("""Each choice should be a 3-tuple: (<value>, <label>, <description>).
+                    value: the actual value in the DB
+                    label: short string to be used in the dropdown when selecting this choice
+                    description: shown after a filter has been chosen, should able to be
+                       inserted in the phrase "Showing voters who <Description>".
+                       E.g. "are <em>female</em>", "live in <em>Orange</em> county".""")
+
+    def get_filter_parms(self) -> Dict:
+        return {self.field_name: self.values[0]}
+
+    def description(self) -> Optional[str]:
+        """
+        Return the appropriate description for the currently selected choice.
+        Or None.
+        """
+        selected_value = self.values[0]
+        for value, label, description in self.choices:
+            if value == selected_value:
+                return description
+
+
+def get_filter_by_name(filter_list, field_name):
+    for filter in filter_list:
+        if filter.field_name == field_name:
+            return filter
+
+
+def filters_from_request(declared_filters: List[Filter], request: HttpRequest) -> Tuple[OrderedDict, Dict]:
+    """
+    Given a list of Filter objects, and an HTTP request:
+    Using the GET parameters from the request, in order, create a new
+    OrderedDict containing copies of the Filter objects corresponding
+    to the query parameter keys, with the values from the query string
+    set in them.
+    Set '.filter_parms' on each new filter to be the cumulative filter parameters.
+    Set '.count' on each new filter to be the count after applying those filters.
+
+    Returns a tuple containing:
+     - an ordered dict with the applied filter objects keyed by field name.
+     - a dict with the final set of filter parameters.
+    """
+    applied_filters = OrderedDict()
+    filter_parms = {}
+
+    for field_name in request.GET:
+        applied_filter = copy(get_filter_by_name(declared_filters, field_name))
```

Comment:
can we rename this something like `filter_inst` because `applied_filter` and `applied_filters` is really easy to mix up trying to read this function?